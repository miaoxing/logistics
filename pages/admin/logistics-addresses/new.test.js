import Page from './new';
import {fireEvent, render, screen} from '@testing-library/react';
import {MemoryRouter} from 'react-router';
import {app} from '@mxjs/app';
import $, {Ret} from 'miaoxing';
import {waitFor} from '@testing-library/dom';
import {bootstrap, createPromise, setUrl, resetUrl} from '@mxjs/test';

bootstrap();

const path = 'admin/logistics-addresses';

describe(path, () => {
  beforeEach(() => {
    setUrl(path + '/new');
    app.page = {
      collection: path,
      index: false,
    };
  });

  afterEach(() => {
    resetUrl();
    app.page = {};
  });

  test('form', async () => {
    const promise = createPromise();
    const promise2 = createPromise();
    const promise3 = createPromise();
    const promise4 = createPromise();
    const promise5 = createPromise();

    $.http = jest.fn()
      // 读取一级地区
      .mockImplementationOnce(() => promise.resolve({
        ret: Ret.suc({
          data: [{
            id: 140000,
            name: '山西省',
            shortName: '山西',
            hasChildren: true,
          }],
        }),
      }))
      // 读取默认数据
      .mockImplementationOnce(() => promise2.resolve({
        ret: Ret.suc({
          data: {
            id: 1,
            name: '联系人1',
            phone: '13800138000',
            region: {
              id: 140222,
              name: '天镇县',
              shortName: '天镇',
              parent: {
                id: 140200,
                name: '大同市',
                shortName: '大同',
                parent: {
                  id: 140000,
                  name: '山西省',
                  shortName: '山西',
                },
              },
            },
            address: '199号',
            types: [1],
            sort: 51,
          },
        }),
      }))
      // 读取二级地区
      .mockImplementationOnce(() => promise3.resolve({
        ret: Ret.suc({
          data: [{
            id: 140200,
            parentId: 140000,
            name: '大同市',
            shortName: '大同',
            hasChildren: true,
          }],
        }),
      }))
      // 读取三级地区
      .mockImplementationOnce(() => promise4.resolve({
        ret: Ret.suc({
          data: [{
            hasChildren: false,
            id: 140222,
            name: '天镇县',
            parentId: 140200,
            shortName: '天镇',
          }],
        }),
      }))
      // 提交
      .mockImplementationOnce(() => promise5.resolve({
        ret: {
          code: 1,
        },
      }));

    const {getByLabelText} = render(<MemoryRouter>
      <Page/>
    </MemoryRouter>);

    await Promise.all([promise, promise2, promise3, promise4]);
    expect($.http).toHaveBeenCalledTimes(4);
    expect($.http).toMatchSnapshot();

    // 看到表单加载了数据
    await waitFor(() => expect(getByLabelText('顺序').value).toBe('51'));
    expect(getByLabelText('联系人姓名').value).toBe('联系人1');

    // 提交表单
    fireEvent.click(screen.getByText('提 交'));

    await Promise.all([promise5]);
    expect($.http).toHaveBeenCalledTimes(5);
    expect($.http).toMatchSnapshot();
  });
});
