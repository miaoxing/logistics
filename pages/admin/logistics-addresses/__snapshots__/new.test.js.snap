// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`admin/logistics-addresses form 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "method": "get",
        "url": "/admin-api/regions?parentId=%E4%B8%AD%E5%9B%BD",
      },
    ],
    Array [
      Object {
        "method": "get",
        "url": "/admin-api/logistics-addresses/defaults",
      },
    ],
    Array [
      Object {
        "method": "get",
        "url": "/admin-api/regions?parentId=140000",
      },
    ],
    Array [
      Object {
        "method": "get",
        "url": "/admin-api/regions?parentId=140200",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
    },
    Object {
      "type": "return",
      "value": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
    },
    Object {
      "type": "return",
      "value": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
    },
    Object {
      "type": "return",
      "value": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
    },
  ],
}
`;

exports[`admin/logistics-addresses form 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "method": "get",
        "url": "/admin-api/regions?parentId=%E4%B8%AD%E5%9B%BD",
      },
    ],
    Array [
      Object {
        "method": "get",
        "url": "/admin-api/logistics-addresses/defaults",
      },
    ],
    Array [
      Object {
        "method": "get",
        "url": "/admin-api/regions?parentId=140000",
      },
    ],
    Array [
      Object {
        "method": "get",
        "url": "/admin-api/regions?parentId=140200",
      },
    ],
    Array [
      Object {
        "data": Object {
          "address": "199号",
          "city": 140222,
          "country": 140000,
          "id": 1,
          "name": "联系人1",
          "phone": "13800138000",
          "postalCode": undefined,
          "province": 140200,
          "regionId": 140222,
          "remark": undefined,
          "sort": 51,
          "types": Array [
            1,
          ],
        },
        "loading": true,
        "method": "POST",
        "url": "/admin-api/logistics-addresses",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
    },
    Object {
      "type": "return",
      "value": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
    },
    Object {
      "type": "return",
      "value": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
    },
    Object {
      "type": "return",
      "value": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
    },
    Object {
      "type": "return",
      "value": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
    },
  ],
}
`;
